juju/model.py

```diff
@@ -2144,6 +2146,91 @@ async def _get_source_api(self, url, controller_name=None):
         await controller.connect(controller_name=controller_name)
         return controller
 
+    async def wait_for_bundle(self, bundle_path, raise_on_error=True,
```

Comment:
This seems reasonable to help tests be better written. I wonder, given its narrow focus - just on a bundle, not a replacement for juju wait etc - whether it should be put in a test helper package initially?

Reply:
I have to agree with @wallyworld here... I think we should iterate this internally in tests for now so that becomes more generic.

Interestingly using the model cache that is backed by the `AllWatcher` will be far superior to how `juju wait` frequently calls the status of juju. Juju status currently requires a database trip to the database for all the data (there are a few exceptions, which hit the internal model cache) and that puts pressure on the database especially when you don't want to put pressure on the database; during a deployment. The `AllWatcher` returns smaller deltas of changes which prevents the mining of superfluous information.

Juju is also [gaining powers](https://github.com/juju/juju/tree/develop/cmd/plugins/juju-wait-for) to wait for certain conditions to be resolved.

Reply:
I might change it to `wait_for_idle` and then move `wait_for_bundle` out to `utils` as a wrapper around `wait_for_idle` which also checks that all of the desired apps and units are there.